Move package versions to ARGs in Dockerfile

thaJeztah opened the pull request:
built on top of https://github.com/moby/moby/pull/40362

- [x] https://github.com/moby/moby/pull/40361 prevent panic if TINI_COMMIT isn't set during build
- [x] https://github.com/moby/moby/pull/40362 Dockerfile: also build old registry on armhf, and small optimization
- [x] https://github.com/moby/moby/pull/40373 Remove e2e dockerfile, and integrate into main Dockerfile

This makes the Dockerfile the cannonical location to specify which version of the dependencies to build.

thaJeztah committed:
Dockerfile: move package versions to ARGs in Dockerfile

This makes the Dockerfile the cannonical location
to specify which version of the dependencies to build.

Also fix some linting issues in the install scripts

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
Dockerfile: preserve TINI_COMMIT build-arg as env

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented on hack/dockerfile/install/tini.installer:
```diff
@@ -1,12 +1,13 @@
 #!/bin/sh
 
-: ${TINI_COMMIT:=fec3683b971d9c3ef73f284f176672c44b448662} # v0.18.0
```
Note; we likely need a make target, or alternative script, because I just discovered that the Docker packaging scripts rely on these; https://github.com/docker/docker-ce-packaging/blob/09b3ac8/deb/common/rules#L12-L15

```
override_dh_auto_build:
	# Build the daemon and dependencies
	cd engine && PRODUCT=docker ./hack/make.sh dynbinary
	cd engine && TMP_GOPATH="/go" hack/dockerfile/install/install.sh tini
	cd engine && TMP_GOPATH="/go" hack/dockerfile/install/install.sh proxy dynamic
	# Build the CLI
```

kolyshkin commented on hack/dockerfile/install/vndr.installer:
```diff
@@ -1,11 +1,12 @@
 	echo "Install vndr version $VNDR_COMMIT"
 	git clone https://github.com/LK4D4/vndr.git "$GOPATH/src/github.com/LK4D4/vndr"
-	cd "$GOPATH/src/github.com/LK4D4/vndr"
+	cd "$GOPATH/src/github.com/LK4D4/vndr" || return
```
There's no need to have this `|| return`, since we do `set -e` in `hack/dockerfile/install/install.sh`.

Same in other places.

thaJeztah replied:
Ah, right; so my linter highlights it, because (obviously) it doesn't know it's being sourced by that file. Perhaps better to repeat the `set -e` in this file, for situations where it's not called through `install.sh` but directly

kolyshkin commented on Dockerfile:
```diff
@@ -51,7 +169,7 @@ RUN --mount=type=cache,target=/root/.cache/go-build \
            go build -buildmode=pie -o /build/registry-v2 github.com/docker/distribution/cmd/registry \
         && case $(dpkg --print-architecture) in \
-               amd64|ppc64*|s390x) \
+               amd64|armhf|ppc64*|s390x) \
```
I can't understand why this is change in this commit

thaJeztah replied:
Moved to https://github.com/moby/moby/pull/40362, and rebased this PR on top of that

kolyshkin commented:
Left a few comments. The last commit is marked as WIP, so marking the whole PR as WIP for now.

thaJeztah committed:
Makefile changes to allow overriding ARGS

Look for environment variables to override versions
in the Dockerfile. This allows, for example:

```
DOCKERCLI_VERSION=19.03.5 \
CONTAINERD_COMMIT=0a1f2b40642e54ed06cd0a22cdf6025cbe70853c \
RUNC_COMMIT=2b52db75279ca687e18156de86d845876e9ef35d \
make shell
```

or

```
CONTAINERD_COMMIT=HEAD RUNC_COMMIT=HEAD make shell
```

or

```
CONTAINERD_COMMIT=master RUNC_COMMIT=master make shell
```

Which produces (inside the container)

```
containerd --version
containerd github.com/containerd/containerd v1.3.0-251-g0a1f2b40 0a1f2b40642e54ed06cd0a22cdf6025cbe70853c

runc --version
runc version 1.0.0-rc9+dev
commit: 2b52db75279ca687e18156de86d845876e9ef35d
spec: 1.0.1-dev

docker --version
Docker version 19.03.5, build 633a0ea
```

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
WIP these vars should be there for hack/make.sh binary

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
Dockerfile: integrate e2e stage in main Dockerfile

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

tao12345666333 commented:
This change is great! Any update on this?
